File: mgyoutube-multitech/api-node/src/repos/simplisticuserdatarepoimpl.ts
import { User, cloneUser } from '../apimodel/user';
import { UserDataRepo } from './userdatarepo';


export class SimplisticUserDataRepoImpl implements UserDataRepo {
    private usernameMap: Map<string, User> = new Map()
    private userIdMap: Map<string, User> = new Map()
    private parentChildrenMap: Map<string, Array<string>> = new Map()
    private nextUserId: number = 1

    constructor() { }

    async repositoryStartup(): Promise<void> {
        // nothing to do here
    }

    async getUserByUsername(username: string): Promise<User | undefined> {
        const user = this.usernameMap.get(username)
        if (user !== undefined) {
            // cloning so modds after return do not affect maps
            return cloneUser(user)
        }
        return user
    }

    async addUser(user: User): Promise<User | undefined> {
        // cloning so that we can mutate userId without affecting the input object
        const addedUser = cloneUser(user)
        addedUser.userId = this.nextUserId.toString();
        this.nextUserId++

        this.userIdMap.set(addedUser.userId, addedUser)
        this.usernameMap.set(addedUser.username, addedUser)

        // cloning so that mods after return do not mutate maps
        // getUserById does the cloning
        return this.getUserById(addedUser.userId);
    }

    async removeUser(user: User): Promise<void> {
        this.userIdMap.delete(user.userId);
        this.usernameMap.delete(user.username);
    }

    async replaceUser(userId: string, user: User): Promise<User | undefined> {

        const existingUser = this.userIdMap.get(userId);

        if (existingUser !== undefined) {
            // cloning so that we can mutate userId without affecting the input object
            const replacingUser = cloneUser(user);
            replacingUser.userId = userId;

            this.userIdMap.set(replacingUser.userId, replacingUser);
            this.usernameMap.set(replacingUser.username, replacingUser);

            // cloning so that mods after return do not mutate maps
            return cloneUser(replacingUser);
        }

        return undefined;
    }


    async addChildToParent(parentUserId: string, childUserId: string): Promise<void> {
        if (!this.parentChildrenMap.has(parentUserId)) {
            this.parentChildrenMap.set(parentUserId, []);
        }

        const childrenUserIds = this.parentChildrenMap.get(parentUserId);

        if (childrenUserIds === undefined) {
            console.error("unable to find child user ides for parent ", parentUserId);
        }
        else if (!childrenUserIds.includes(childUserId)) {
            childrenUserIds.push(childUserId);
        }
    }

    async getChildrenForParent(parentUserId: string): Promise<Array<User>> {
        const children: Array<User> = []

        if (this.parentChildrenMap.has(parentUserId)) {
            const childrenUserIds = this.parentChildrenMap.get(parentUserId);

            if (childrenUserIds !== undefined) {
                for (var childUserId of childrenUserIds) {
                    const user = await this.getUserById(childUserId);
                    if (user !== undefined) {
                        children.push(user);
                    } else {
                        console.warn("unknown child userid ", childUserId);
                    }
                }
            }
        }

        return children;
    }

    // probably will be public eventually
    async getUserById(userId: string): Promise<User | undefined> {
        const foundUser = this.userIdMap.get(userId);
        if (foundUser === undefined) {
            return undefined;
        }
        return cloneUser(foundUser);
    }
}

File: mgyoutube-multitech/api-python/repos/SimplisticUserDataRepoImpl.py
from repos.UserDataRepo import UserDataRepo
from apimodel.User import User, cloneUser


class SimplisticUserDataRepoImpl(UserDataRepo):

    def __init__(self):
        self.usernameMap = dict()
        self.userIdMap = dict()
        self.parentChildrenMap = dict()
        self.nextUserId = 1

    def getUserByUsername(self, username: str) -> User:
        user = self.usernameMap.get(username)
        if (user != None):
            # cloning so modds after return do not affect maps
            return cloneUser(user)
        return user

    def addUser(self, user: User) -> User:
        # cloning so that we can mutate userId without affecting the input object
        addedUser = cloneUser(user)
        addedUser.userId = str(self.nextUserId)
        self.nextUserId = self.nextUserId + 1

        print("SimplisticUserDataRepoImpl.addUser, user being added",
              addedUser, "nextUserId", self.nextUserId)

        self.userIdMap[addedUser.userId] = addedUser
        self.usernameMap[addedUser.username] = addedUser

        # cloning so that mods after return do not mutate maps
        return cloneUser(addedUser)

    def removeUser(self, user: User):
        del self.userIdMap[user.userId]
        del self.usernameMap[user.username]

    def replaceUser(self, userId: str, user: User) -> User:
        existingUser = self.userIdMap[userId]

        if existingUser is not None:
            # // cloning so that we can mutate userId without affecting the input object
            replacingUser = cloneUser(user)
            replacingUser.userId = userId

            self.userIdMap[replacingUser.userId] = replacingUser
            self.usernameMap[replacingUser.username] = replacingUser

            # // cloning so that mods after return do not mutate maps
            return cloneUser(replacingUser)

        return None

    def addChildToParent(self, parentUserId: str, childUserId: int):
        childList = self.parentChildrenMap.get(parentUserId)
        if childList is None:
            self.parentChildrenMap[parentUserId] = list()

        childrenUserIds = self.parentChildrenMap[parentUserId]

        try:
            childrenUserIds.index(childUserId)
        except ValueError:
            childrenUserIds.append(childUserId)

    def getChildrenForParent(self, parentUserId: str) -> list:
        children = list()

        childrenUserIds = self.parentChildrenMap.get(parentUserId)
        if childrenUserIds is not None:
            for childUserId in childrenUserIds:
                user = self.getUserById(childUserId)
                if user is not None:
                    children.append(user)
                else:
                    print("unknown child userid ", childUserId)

        return children

    def getUserById(self, userId: str) -> User:
        return self.userIdMap.get(userId)
